### PR TITLE
Make python test suites run in parallel

### PR DESCRIPTION
**Pull request by jtattermusch**
-- allow filtering test cases using a prefix in _runner.py
-- added tests.json file that contains list of all existing python test suites (the list is used by run_tests.py to determine which suites to run). To make sure that the tests.json file is always up-to-date, I added a "sanity" tests that discovers all existings tests using test loader and checks that tests.json is up-to-date.
-- gcov still runs all the tests serially and should have the same behavior as before

The work is not 100% finished yet, but would be worth reviewing.

### CODE REVIEW

**Comment by jtattermusch** [comment]
@soltanmm  please also take a look.

**Comment by jtattermusch** [comment]
@ctiller  @nicolasnoble  FYI

File: tools/run_tests/run_python.sh [hunk 1]
```diff
@@ -42,7 +42,12 @@ export LDFLAGS="-L$ROOT/libs/$CONFIG"
 export GRPC_PYTHON_BUILD_WITH_CYTHON=1
 export GRPC_PYTHON_ENABLE_CYTHON_TRACING=1
 
-tox
+if [ "$CONFIG" = "gcov" ]
```
Comment (soltanmm) [inline tools/run_tests/run_python.sh]: I'm confused. What's this branch for?

Comment (jtattermusch) [inline tools/run_tests/run_python.sh]: for code coverage we use the commands from tox.ini to run the tests and generate the coverage report.
For a "regular test run" (=opt or dbg), we are running tests in parallel and tox cannot handle that. So we make sure everything is built in advance and just run the test here.

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import unittest
+
+
+class Sanity(unittest.TestCase):
+
+  def testTestsJsonUpToDate(self):
+    """Autodiscovers all test suites and checks that tests.json is up to date"""
+    loader = _loader.Loader()
+    loader.loadTestsFromNames(['tests'])
+    test_suite_names = [c.id().rsplit('.', 1)[0]
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: Please avoid single-letter identifiers.

The values assigned to `c` are `unittest.TestCase` subclasses, right? How about `test_case_class` for a name? You can win some horizontal space by line-wrapping immediately after `[`:
```
    test_suite_names = [
        test_case_class.id().rsplit('.', 1)[0]
        for test_case_class in _loader.iterate_suite_cases(loader.suite)]
```
.

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                        for c in _loader.iterate_suite_cases(loader.suite)]
+    test_suite_names = sorted(set(test_suite_names))
+
+    with open('src/python/grpcio/tests/tests.json') as f:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: `tests_json_file` rather than `f`, please.

File: tools/run_tests/run_tests.py [hunk 4]
```diff
@@ -353,15 +353,27 @@ def configure(self, config, args):
     _check_compiler(self.args.compiler, ['default'])
 
   def test_specs(self):
+    # load list of known test suites
+    with open('src/python/grpcio/tests/tests.json') as f:
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: `tests_json_file` here too?

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+from tests import _loader
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: Standard-library imports come first, then local imports. Also you should import any underscore-prefixed attribute of another code element.

I think these imports should be
```
import json
import unittest

import tests
```
and then below where you currently have `_loader.Loader()` you would just have `tests.Loader()`.

**Comment by nathanielmanistaatgoogle** [comment]
Impressive; I like what you've done here.

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py (hunk 5 above)
Comment (jtattermusch) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: done

File: tools/run_tests/run_tests.py (hunk 4 above)
Comment (jtattermusch) [inline tools/run_tests/run_tests.py]: done

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py (hunk 3 above)
Comment (jtattermusch) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: done

File: src/python/grpcio/tests/unit/_sanity/_sanity_test.py (hunk 2 above)
Comment (jtattermusch) [inline src/python/grpcio/tests/unit/_sanity/_sanity_test.py]: done

**Comment by jtattermusch** [comment]
Ok, I addressed all the comments and squashed the commits. Let's see how the test results look like.
Also, thanks for a little lesson of good python style.

**Comment by nathanielmanistaatgoogle** [comment]
Commit content looks good, commit itself looks good, Python is green, sanity is [`test/cpp/interop/reconnect_interop_client.cc modified 2016 but copyright only extends to 2015`](https://en.wikipedia.org/wiki/Somebody_else's_problem). I'm satisfied with this but @soltanmm will also review and approve before merge.

**Comment by soltanmm** [comment]
:thumbsup:

I'm fine with this for now.

**Pull request merged**